drivers/dose/dose.c

```diff
@@ -0,0 +1,557 @@
+        DEBUG("dose _recv(): receive buffer too small\n");
+        clear_recv_buf(ctx);
+        return -1;
+    }
+    else {
+        /* Copy the packet to the provided buffer. */
+        memcpy(buf, ctx->recv_buf, pktlen);
+        clear_recv_buf(ctx);
+        return pktlen;
+    }
+}
+
+static uint8_t wait_for_state(dose_t *ctx, uint8_t state)
+{
+    do {
+        /* This mutex is unlocked by the state machine
+         * after every state transition */
+        mutex_lock(&ctx->state_mtx);
+    } while (state != DOSE_STATE_ANY && ctx->state != state);
+    return ctx->state;
+}
+
+static int send_octet(dose_t *ctx, uint8_t c)
+{
+    uart_write(ctx->uart, (uint8_t *) &c, sizeof(c));
+
+    /* Wait for a state transition */
+    uint8_t state = wait_for_state(ctx, DOSE_STATE_ANY);
+    if (state != DOSE_STATE_SEND) {
+        /* Timeout */
+        DEBUG("dose send_octet(): timeout\n");
+        return -2;
+    }
+    else if (ctx->uart_octet != c) {
+        /* Mismatch */
+        DEBUG("dose send_octet(): mismatch\n");
+        return -1;
+    }
+
+    return 0;
+}
+
+static int send_data_octet(dose_t *ctx, uint8_t c)
+{
+    int rc;
+
+    /* Escape special octets */
+    if (c == DOSE_OCTECT_ESC || c == DOSE_OCTECT_END) {
+        rc = send_octet(ctx, DOSE_OCTECT_ESC);
+        if (rc) {
+            return rc;
+        }
+    }
+
+    /* Send data octet */
+    rc = send_octet(ctx, c);
+
+    return rc;
+}
+
+static int _send(netdev_t *dev, const iolist_t *iolist)
+{
+    dose_t *ctx = (dose_t *) dev;
+    int8_t retries = 3;
+    size_t pktlen;
+    uint16_t crc;
+
+send:
+    crc = 0xffff;
+    pktlen = 0;
+
+    /* Switch to state SEND */
+    do {
+        wait_for_state(ctx, DOSE_STATE_IDLE);
+        state(ctx, DOSE_SIGNAL_SEND);
+    } while (wait_for_state(ctx, DOSE_STATE_ANY) != DOSE_STATE_SEND);
+
+    /* Send packet buffer */
+    for (const iolist_t *iol = iolist; iol; iol = iol->iol_next) {
+        size_t n = iol->iol_len;
+        pktlen += n;
+        uint8_t *ptr = iol->iol_base;
+        while (n--) {
+            /* Send data octet */
+            if (send_data_octet(ctx, *ptr)) {
+                goto collision;
+            }
+
+            /* Update CRC */
+            crc = crc16_update(crc, *ptr);
+
+            ptr++;
+        }
+    }
+
+    /* Send CRC */
+    network_uint16_t crc_nw = byteorder_htons(crc);
+    if (send_data_octet(ctx, crc_nw.u8[0])) {
+        goto collision;
+    }
+    if (send_data_octet(ctx, crc_nw.u8[1])) {
+        goto collision;
+    }
+
+    /* Send END octet */
+    if (send_octet(ctx, DOSE_OCTECT_END)) {
+        goto collision;
+    }
+
+    /* We probably sent the whole packet?! */
+    ctx->netdev.event_callback((netdev_t *) ctx, NETDEV_EVENT_TX_COMPLETE);
+
+    /* Get out of the SEND state */
+    state(ctx, DOSE_SIGNAL_END);
+
+    return pktlen;
+
+collision:
+    DEBUG("dose _send(): collision!\n");
+    if (--retries < 0) {
+        return 0;
+    }
+    goto send;
```

Comment:
Call `ctx->netdev.event_callback((netdev_t *) ctx, NETDEV_EVENT_TX_MEDIUM_BUSY);` if we weren't able to send the packet. This will increase the `tx_failed` counter for the `ifconfig` command.

Reply:
Added with commit 5959328